Add verify_eigrp_router_id API to verify.py  ios/iosxe eigrp protocol (my old contribution) supporting EIGRPv4 and EIGRPv6 for IOS/IOSXE

cherifimehdi opened the pull request:
Hello,

I added verify_eigrp_router_id API to verify.py (my old contribution) for ios/iosxe eigrp protocol.

I added the tests at pkgs/sdk-pkg/src/genie/libs/sdk/apis/tests/iosxe/eigrp/verify

I added the changelog_add_verify_eigrp_router_id_202421031402 file at pkgs/sdk-pkg/changelog folder

Thank you for reviewing my contribution

cherifimehdi commented:
Hello @omehrabi @lsheikal thank you for reviewing my PR.
Thanks

cherifimehdi commented:
Hi @lsheikal  @omehrabi please can you review and merge my PR since it is an addition to my previous contribution (verifi.py for iosxe eigrp) and I followed same logic based on reviewers from my previous PR @ThomasJRyan and @Lelor.
Thank you

cherifimehdi commented:
Hi @omehrabi  I fixed the problem with the update tests after 24.3 release

cherifimehdi commented:
> there is a change in unicon.plugin which will break the unittest for genielibs please update tests after 24.3 release

Thank you, Hi @omehrabi , the problem with the update is fixed

lsheikal commented on pkgs/sdk-pkg/src/genie/libs/sdk/apis/iosxe/eigrp/verify.py:
```diff
@@ -122,3 +122,62 @@ def verify_eigrp_neighbors(
+    """verify EIGRP IDs for a given vrf and AS active instance for ipv4 or ipv6
+    Args:
+      device (obj): Device object
+      router_id = None (list): List of router ID to check
+      vrf = "default" (str): Name of the vrf by default set to "default"
+      AS_N = None (int) : Autonomous System
+      ip = "ipv4" (str): Protocol ip set by default to "ipv4" to change to "ipv6"
```
```suggestion
      router_id  (list): List of router ID to check
      vrf  (str): Name of the vrf by default set to "default"
      AS_N (int) : Autonomous System
      ip (str): Protocol ip set by default to "ipv4" to change to "ipv6"
```

cherifimehdi replied:
Hi @lsheikal I think here is the same. there is no change to make!

lsheikal replied:
it is a required change, to keep the docstrings consistent.

lsheikal commented on pkgs/sdk-pkg/src/genie/libs/sdk/apis/iosxe/eigrp/verify.py:
```diff
@@ -122,3 +122,62 @@ def verify_eigrp_neighbors(
+    timeout = Timeout(max_time, check_interval)
+    while timeout.iterate():
+        if AS_N and router_id:
+
```
removve extra line.

cherifimehdi replied:
Hi, @lsheikal  I fixed this

lsheikal commented on pkgs/sdk-pkg/src/genie/libs/sdk/apis/iosxe/eigrp/verify.py:
```diff
@@ -122,3 +122,62 @@ def verify_eigrp_neighbors(
+      device (obj): Device object
+      router_id = None (list): List of router ID to check
+      vrf = "default" (str): Name of the vrf by default set to "default"
+      AS_N = None (int) : Autonomous System
```
need a better naming(uppercase not allowed). how about auto_sys instead of AS_N?

cherifimehdi replied:
Hi @lsheikal here I prefer (if it's possible!) to leave AS_N since all the APIs I added in get.py and verify.py contain the same variable AS_N

cherifimehdi replied:
@lsheikal I committed the suggestions

lsheikal replied:
We follow PEP8 standards, its not allowed. Kindly change in other apis too.
https://peps.python.org/pep-0008/#function-and-method-arguments

lsheikal commented on pkgs/sdk-pkg/src/genie/libs/sdk/apis/iosxe/eigrp/verify.py:
```diff
@@ -122,3 +122,60 @@ def verify_eigrp_neighbors(
+      True
+      False
+    """
+    assert isinstance(AS_N, int), "AS_N must be int"
+    assert isinstance(vrf, str), "vrf must be str"
+    assert isinstance(router_id, list), "router_id must be list of router IDs"
+    assert AS_N != 0, "AS_N must not be 0"
+    assert ip in ["ipv4", "ipv6"], "ip must be ipv4 or ipv6"
```
comment before assertion.

cherifimehdi replied:
Hi @lsheikal  There is a comment just after the comma, here an assertion with comment for example assert isinstance(AS_N, int), "AS_N must be int" the comment is ""AS_N must be int""

lsheikal commented on pkgs/sdk-pkg/src/genie/libs/sdk/apis/iosxe/eigrp/verify.py:
```diff
@@ -122,3 +122,60 @@ def verify_eigrp_neighbors(
+    assert isinstance(router_id, list), "router_id must be list of router IDs"
+    assert AS_N != 0, "AS_N must not be 0"
+    assert ip in ["ipv4", "ipv6"], "ip must be ipv4 or ipv6"
+    timeout = Timeout(max_time, check_interval)
```
why this timeout needed? comment.

cherifimehdi replied:
Timeout is needed to wait before continue for the parsed data

lsheikal commented on pkgs/sdk-pkg/src/genie/libs/sdk/apis/iosxe/eigrp/verify.py:
```diff
@@ -122,3 +122,60 @@ def verify_eigrp_neighbors(
+                    .contains((ip[0:2]).upper() + ip[2:])
+                    .get_values("eigrp_id")
+                )
+            return set(router_id).issubset(eigrp_id)
```
comment an example set with a subset.

lsheikal replied:
and explain what it verifies.

cherifimehdi commented on pkgs/sdk-pkg/src/genie/libs/sdk/apis/iosxe/eigrp/verify.py:
```diff
@@ -122,3 +122,60 @@ def verify_eigrp_neighbors(
+                    .contains((ip[0:2]).upper() + ip[2:])
+                    .get_values("eigrp_id")
+                )
+            return set(router_id).issubset(eigrp_id)
```
Verify router id if exist as the Name of the API indicate

cherifimehdi commented:
Hi @lsheikal why there is problem with test?

lsheikal commented:
> Hi @lsheikal why there is problem with test?

pull the latest dev and merge with your branch. And address my comments.

cherifimehdi reviewed (commented):
Fixed

cherifimehdi reviewed (commented):
Change applied

cherifimehdi commented:
Hi @lsheikal same problem, I don't know way test failed

cherifimehdi commented:
Hi @omehrabi  @lsheikal  I think problem with test fixed

cherifimehdi commented:
Hi @omehrabi  @lsheikal  are the corrections I made OK or There are modifications to do? Thank you

cherifimehdi commented:
> > Hi @omehrabi @lsheikal are the corrections I made OK or There are modifications to do? Thank you
> 
> you still have not addressed the comments, i can still see the parameter `AS_N ` in upper case, which violates pep8 standards. change it to a meaningful parameter name in lowercase.

Hi @omehrabi  @lsheikal thank you. should I modify 'AS_N' just for this API or for all previous APIs merged previously

lsheikal commented:
> > > Hi @omehrabi @lsheikal are the corrections I made OK or There are modifications to do? Thank you
> > 
> > 
> > you still have not addressed the comments, i can still see the parameter `AS_N ` in upper case, which violates pep8 standards. change it to a meaningful parameter name in lowercase.
> 
> Hi @omehrabi @lsheikal thank you. should I modify 'AS_N' just for this API or for all previous APIs merged previously

lets modify this parameter for this api. And kindly adhere to the pep8 standards hereafter.

cherifimehdi commented:
> > > > Hi @omehrabi @lsheikal are the corrections I made OK or There are modifications to do? Thank you
> > > 
> > > 
> > > you still have not addressed the comments, i can still see the parameter `AS_N ` in upper case, which violates pep8 standards. change it to a meaningful parameter name in lowercase.
> > 
> > 
> > Hi @omehrabi @lsheikal thank you. should I modify 'AS_N' just for this API or for all previous APIs merged previously
> 
> lets modify this parameter for this api. And kindly adhere to the pep8 standards hereafter.

Hi @omehrabi  @lsheikal  I changed 'AS_N' to 'auto_sys' in the API

cherifimehdi commented:
Hi @lsheikal, are the changes OK or are there more to do besides "AS_N" (I change it as requested)? THANKS

omehrabi reviewed (changes requested):
there have been a change in unicon please update to the latest version and record the mock data

cherifimehdi commented:
Hi @omehrabi  I updated to the latest version and I updated mock data and tests. Thank you

cherifimehdi commented:
Hi  @omehrabi  Is there any other change to make or it is ok? Thanks
